src/pac.c

```diff
@@ -158,12 +167,15 @@ pac_alloc_impl(tsdn_t *tsdn, pai_t *self, size_t size, size_t alignment,
 	pac_t *pac = (pac_t *)self;
 	ehooks_t *ehooks = pac_ehooks_get(pac);
 
-	edata_t *edata = pac_alloc_real(tsdn, pac, ehooks, size, alignment,
-	    zero, guarded);
+	edata_t *edata = NULL;
+	if (!guarded || frequent_reuse) {
```

Comment:
Is it true that the `frequent_reuse` condition is an optimization? i.e. if it's a guarded && large request, going down pac_alloc_real will return NULL gracefully, because we never put large back into the guarded ecache (and won't attempt to grow retained)?

If so, some comments here can certainly help.

Reply:
Sounds good. Maybe it's slightly easier to read with

    bool try_cached_guarded = guarded && frequent_reuse;
    if (!try_cached_guarded) { ...

Reply:
Well, maybe we simply don't do this optimization? By default, slab sizes can be 5 and 7 pages (which are large sizes), and that mean even if frequent_reuse == false, we may still find cached & guarded extents. We can also say that we don't really want those cached extents to be used that way, but it doesn't hurt much.

Reply:
I've added a comment. Let me know if it makes it clear what's going on.